Add edit book page

phucla0122 opened the pull request:
- Completed the add and edit features of the bookstore

- Added test cases for both features
- close #4

phucla0122 committed:
Create add/edit page, still need some testing

phucla0122 committed:
Create add/edit page, still need some testing

phucla0122 committed:
Created add book, work well

phucla0122 committed:
Created add book, work well

phucla0122 committed:
Completed book edit and test case for add and edit book

nicolastuttle commented on src/main/java/com/groupseven/sysc4806project/BookstoreController.java:
```diff
@@ -39,4 +42,53 @@ public String home(
     }
+
+    @PostMapping("/add-book")
+    public String add(
```
The params for this method should probably follow the same rules as `BookService.create()`. ie: ISBN, title, author, and publisher are required, inventory defaults to 0. We can let it error if there's missing params, maybe put an error message on the add page

phucla0122 replied:
I changed this so that it would check for the possible to be missing attribute

nicolastuttle replied:
The concern I have with this is it allows someone to add a book with no title/isbn/author/publisher. Could just have a blank entry

phucla0122 replied:
I see, alright. I'll check if it's blank or null then

nicolastuttle commented on src/main/resources/templates/add-edit.html:
```diff
@@ -0,0 +1,37 @@
+    <form th:unless="${book == null}" action="/home/edit-book" method="POST">
+      <p>ISBN: <input type="text" name="isbn" class="border-2 rounded-md" th:value="${book.getIsbn()}" disabled/>
+      <p>Title: <input type="text" name="title" class="border-2 rounded-md" th:value="${book.getTitle()}" disabled/>
+      <p>Description: <input type="text" name="description" class="border-2 rounded-md"/>
```
Can we have this filled out with the current description as a placeholder? Also this input might be better as a `<textarea>`, more space

phucla0122 replied:
I did change the input to text area for the description. As for the placeholder, what do you mean by that?

nicolastuttle replied:
Just set `th:value="${book.getDescription()}"` so if, for example, the bookstore owner wanted to add a sentence to the description or fix a typo they don't have to rewrite the entire description

phucla0122 replied:
Alright, will do that then

nicolastuttle replied:
Looks like you changed the book ID input to be a textarea which broke the edit screen. You'll have to switch that back to being an input and make the description input the textarea instead

phucla0122 replied:
My bad on that one, was kinda sleepy when I made the changed. Misread the ID for the description. I fixed it

nicolastuttle commented on src/main/resources/templates/add-edit.html:
```diff
@@ -0,0 +1,37 @@
+      <p>Description: <input type="text" name="description" class="border-2 rounded-md"/>
+      <p>Author: <input type="text" name="author" class="border-2 rounded-md" th:value="${book.getAuthor()}" disabled/>
+      <p>Publisher: <input type="text" name="publisher" class="border-2 rounded-md" th:value="${book.getPublisher()}" disabled/>
+      <p>Inventory: <input type="number" name="inventory" class="border-2 rounded-md"/>
```
Would be nice to see current inventory as placeholder

phucla0122 replied:
Same question as above

nicolastuttle commented on src/main/resources/templates/add-edit.html:
```diff
@@ -0,0 +1,37 @@
+  <script src="https://cdn.tailwindcss.com"></script>
+</head>
+<body>
+  <div>
```
I think you can collapse both forms into one using conditionals in your `th:value` options.

ex: `<input type="text" name="isbn" class="whatever" th:value="${book == null ? '' : book.getIsbn()}" th:disabled="${book == null}">`

phucla0122 replied:
I try to do so, but there are some problems with the disabled field for it. That field is conditional diable, but it's the opposite of the book == null condition. We want it to be not disabled when the book is null. But I don't know how to change that in the th:disabled. Also, thymeleaf don't recognize book != null for some odd reason.

nicolastuttle replied:
It doesn't work with `th:disabled`?

phucla0122 replied:
from how i try it, it seems like so

nicolastuttle replied:
I've gotten it to work locally, are you sure you're putting the right condition? `th:disabled="${book != null}"` or try it with the inverse condition to see if you can edit the "locked" book fields

nicolastuttle replied:
I've gotten it to work locally, are you sure you're putting the right condition? `th:disabled="${book != null}"` or try it with the inverse condition to see if you can edit the "locked" book fields

phucla0122 replied:
I've changed it and now it's working

nicolastuttle commented on src/main/resources/templates/add-edit.html:
```diff
@@ -0,0 +1,37 @@
+      <button class="ease-in-out duration-300 hover:bg-blue-800 hover:text-white border-blue-800 border-2 rounded-md px-4 py-1" type="submit">Add</button>
+      <button class="ease-in-out duration-300 hover:bg-blue-800 hover:text-white border-blue-800 border-2 rounded-md px-4 py-1" type="Reset">Reset</button>
+    </form>
+
```
Maybe we could have a "back" button too just to send us back to the home page?

phucla0122 replied:
Back button added

robellgabriel commented on src/main/resources/templates/add-edit.html:
```diff
@@ -0,0 +1,37 @@
+      <p>Description: <input type="text" name="description" class="border-2 rounded-md"/>
+      <p>Author: <input type="text" name="author" class="border-2 rounded-md" th:value="${book.getAuthor()}" disabled/>
+      <p>Publisher: <input type="text" name="publisher" class="border-2 rounded-md" th:value="${book.getPublisher()}" disabled/>
+      <p>Inventory: <input type="number" name="inventory" class="border-2 rounded-md"/>
```
You can set the number to be >0 with adding `min=0`

phucla0122 replied:
Thanks for the suggestion

robellgabriel commented on src/main/java/com/groupseven/sysc4806project/BookstoreController.java:
```diff
@@ -39,4 +42,53 @@ public String home(
+            inventory = 0;
+        }
+        this.bookService.update(book_id,isbn,title,description,author,publisher,inventory,image);
+        return "redirect:/home";
```
just a question, is redirect necessary? would it still work with using `return "index"`

nicolastuttle replied:
`return "index"` would require populating the Model for the page again. Using the redirect allows us to reuse the function to render the homepage

phucla0122 replied:
not quite since if you just return index, it would just return the page without any book on it since /home is the thing that renders the book to index.

robellgabriel replied:
ah ok I see, will definitley use this for the cart page ty!

robellgabriel commented on src/main/resources/templates/add-edit.html:
```diff
@@ -0,0 +1,37 @@
+      <p>Description: <input type="text" name="description" class="border-2 rounded-md"/>
+      <p>Author: <input type="text" name="author" class="border-2 rounded-md"/>
+      <p>Publisher: <input type="text" name="publisher" class="border-2 rounded-md"/>
+      <p>Inventory: <input type="number" name="inventory" class="border-2 rounded-md"/>
```
same as prev comment

phucla0122 replied:
Yep, and i fixed it

phucla0122 committed:
Merge branch 'add_edit_book_page' of https://github.com/CarletonU-2223-SAV/SYSC4806-Project into add_edit_book_page

phucla0122 committed:
Changed some part for the add-book sesrvice, edit some changes to the add-edit html regarding the rendering of the page

phucla0122 committed:
Adding placeholder for book's description in edit mode

nicolastuttle commented on src/main/java/com/groupseven/sysc4806project/BookstoreController.java:
```diff
@@ -39,4 +42,63 @@ public String home(
+        if (inventory < 0) {
+            inventory = 0;
+        }
+        if(isbn == null) {
```
These checks will never be true as Spring will throw an exception if these params are missing before we get here

phucla0122 replied:
Then I don't think we need it anyway, since spring will check for the missing parameters

nicolastuttle replied:
I agree. If the checks are not required we should take them out

phucla0122 replied:
Alright, fair enough

nicolastuttle commented on src/main/resources/templates/index.html:
```diff
@@ -77,6 +79,10 @@
                             <button class="ease-in-out duration-300 hover:bg-blue-800 hover:text-white border-blue-800 border-2 rounded-md px-4 py-1 ml-2">Add to cart</button>
+                            <form action="/home/transit-to-edit-book" method="GET">
+                                <input name="id" type="hidden"th:value="${book.getId()}">
+                                <button class="ease-in-out duration-300 hover:bg-blue-800 hover:text-white border-blue-800 border-2 rounded-md px-4 py-1 ml-2">Edit book</button>
```
While you're in here could you add a delete book button as well please?

phucla0122 replied:
I added the delete button, but I made a couple of changing to the mapping due to thymeleaf not supporting PUT and DELETE method. Overall, I won't affect the program

phucla0122 committed:
Merge branch 'master' into add_edit_book_page

phucla0122 committed:
Many changes to the program

- Added delete button to the book
- Change mapping for the book as thymeleaf does not support DELETE and PUT http methods.
- Clean up add-edit.html to make it more concise

phucla0122 committed:
Minor changes to test unit as api changes for delete

phucla0122 committed:
remove checking for add book api
